### PR TITLE
Fees for Ethereum → Sustrate messages

### PR DESCRIPTION
**Pull request by vgeddes**
This change implements fees for sending messages to Sustrate as well as a bunch of related things:

* In Substrate, added a Treasury pallet
* The fee attached to an incoming message is split, 80% going to the relayer account, and 20% going to the treasury. Of course this can be changed as necessary.
* On Ethereum, users need to authorize apps to send messages on their behalf. Our current set of bridge apps however are authorized by default for all users. See ChannelAccess.sol for this code which is used by IncentivizedOutboundChannel.
* On Ethereum, initial governance / configuration mechanisms for some contracts.

The changes have all been tested using the updated E2E and unit tests.

### CODE REVIEW

File: ethereum/contracts/IncentivizedOutboundChannel.sol
Comment (musnit) [inline ethereum/contracts/IncentivizedOutboundChannel.sol:60]: can use a modifier for this like onlyConfigUpdater - https://docs.soliditylang.org/en/v0.6.2/contracts.html#function-modifiers

File: ethereum/migrations/2_next.js
Comment (musnit) [inline ethereum/migrations/2_next.js:34]: put in config plz

File: ethereum/contracts/ChannelAccess.sol
Comment (musnit) [inline ethereum/contracts/ChannelAccess.sol:5]: Can we get a high level comment describing the problem being solved by this as a guide to readers/auditors?

File: ethereum/migrations/2_next.js
Comment (musnit) [inline ethereum/migrations/2_next.js:105]: put in config plz

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (musnit) [inline parachain/pallets/incentivized-channel/src/inbound/mod.rs:119]: @philipstanislaus plz review this function thoroughly too

**Review by musnit — APPROVED** [review]
some minor comments, but otherwise looks good. lets get another approval before merging though.

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:119]: Just to explain this function a bit more (and I'll add a comment to the code as well):

* If the fee is zero, do nothing
* Otherwise, withdraw the fee amount from the Dot-App module account, returning a negative imbalance
* Figure out the fraction of the fee amount that should be paid to the relayer
* Pay the relayer if their account exists.
* Adjust the negative imbalance by offsetting the amount paid to the relayer
* Resolve the negative imbalance by depositing it into the treasury account

Here's a primer on imbalances to explain that aspect a bit more: https://blog.polymath.network/substrate-deep-dive-imbalances-8dfa89cc1d1

File: parachain/pallets/incentivized-channel/src/inbound/test.rs
Comment (Rizziepit) [inline parachain/pallets/incentivized-channel/src/inbound/test.rs:271]: You use `resolve_creating` in `handle_fee`. So it should create this account if it doesn't exist, right? I'd add a test for that

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (Rizziepit) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:119]: This looks fine to me. Perhaps remove the `RewardRelayer` stuff from core though

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (Rizziepit) [inline parachain/pallets/incentivized-channel/src/inbound/mod.rs:127]: Heads up that `debug` becomes `log` in https://github.com/Snowfork/polkadot-ethereum/pull/331

**Review by Rizziepit — COMMENTED** [review]
I reviewed the Substrate code. lgtm

File: test/test/test.js
Comment (philipstanislaus) [inline test/test/test.js:54]: Is there a better way to wait for the event arriving on Ethereum than sleeping?

**Review by philipstanislaus — APPROVED** [review]
LGTM, few minor comments

File: ethereum/contracts/ChannelAccess.sol
Comment (vgeddes) [reply ethereum/contracts/ChannelAccess.sol:26]: I adapted this from the ERC-777 [operators](https://eips.ethereum.org/EIPS/eip-777#operators-1) feature, so I mostly went with their design.

Can always change this if required. But for now its fine if only users and governance contracts can revoke.

File: ethereum/contracts/ChannelAccess.sol
Comment (vgeddes) [reply ethereum/contracts/ChannelAccess.sol:18]: added docs in latest upload.

File: ethereum/migrations/2_next.js
Comment (vgeddes) [reply ethereum/migrations/2_next.js:34]: That's actually just stub code until I develop the governance contract and deploy it. No need for config

File: ethereum/contracts/IncentivizedOutboundChannel.sol
Comment (vgeddes) [reply ethereum/contracts/IncentivizedOutboundChannel.sol:60]: Will keep it mind when I flesh out governance on the Ethereum side.

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:132]: Nah, [Perbill](https://substrate.dev/rustdocs/v3.0.0/sp_runtime/struct.Perbill.html) is always a fraction in the range `[0, 1]`.

Comment (philipstanislaus) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:132]: Okay, that‘s good to know!

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:124]: I don't think that actually has any effect at the moment for the balances pallet, looking at https://github.com/paritytech/substrate/blob/master/frame/balances/src/lib.rs#L532

The only reason that seems to change any withdrawal logic is `TRANSACTION_PAYMENT`. Presumably other implementations of `Currency` may do something different.

In any case, `T::SourceAccount` actually maps to the DOTApp account, rather than the account of a specific user. So I don't see any semantic benefit in marking the withdrawal as a fee. We're not really charging a fee from the DOTAccount in exchange for some action.

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:138]: `rewarded` will always be less than `imbalance`, ignoring negative vs positive.

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:146]: Ah, `adjusted_imbalance` is always a `NegativeImbalance`. It will always compile.

File: test/test/test.js
Comment (vgeddes) [reply test/test/test.js:54]: Definitely, we can always update the E2E tests to watch for events. Its just very low-hanging fruit in the scheme of things.

File: parachain/pallets/incentivized-channel/src/inbound/test.rs
Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/test.rs:281]: Yeah the E2E tests do verify that the treasury account is created and has the right amount deposited.

File: parachain/pallets/incentivized-channel/src/inbound/mod.rs
Comment (philipstanislaus) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:133]: I think the main difference is how it will show up in block explorer, if at all. A transfer has a sender and a recipient, while a fee is not "sent" to anyone. This is hypothetical though, because block explorers currently use events rather than withdraw reasons. 

No strong opinion, but to not mess up explorers, I guess changing to `FEE` might make sense.

Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:133]: Hey yeah I remember we previously chatted about that. I agree. That's why I moved to withdraw & deposit instead of transfer, as events are only emitted for the latter (because its a high level operation). So we're fine in that regard.

Comment (vgeddes) [reply parachain/pallets/incentivized-channel/src/inbound/mod.rs:133]: For reference, no events are emitted in `Currency::withdraw`: https://github.com/paritytech/substrate/blob/master/frame/balances/src/lib.rs#L1473

**Pull request merged**
